scico/flax/train/trainer.py

```diff
@@ -0,0 +1,542 @@
+        self.steps_per_epoch: int = len_train // batch_size
+        config["steps_per_epoch"] = self.steps_per_epoch  # needed for creating lr schedule
+        self.num_steps: int = int(self.steps_per_epoch * num_epochs)
+
+        # Evaluation (over testing set) steps
+        num_validation_examples: int = len_test
+        if "steps_per_eval" not in config:
+            self.steps_per_eval: int = num_validation_examples // batch_size
+        else:
+            self.steps_per_eval = config["steps_per_eval"]
+
+        # Determine monitoring steps
+        if "steps_per_checkpoint" not in config:
+            self.steps_per_checkpoint: int = self.steps_per_epoch * 10
+        else:
+            self.steps_per_checkpoint = config["steps_per_checkpoint"]
+
+        if "log_every_steps" not in config:
+            self.log_every_steps: int = self.steps_per_epoch * 20
+        else:
+            self.log_every_steps = config["log_every_steps"]
+
+    def configure_reporting(self, config: ConfigDict):
+        """Configure logging and checkpointing.
+
+        The parameters configured correspond to
+
+        - log: A flag for logging to the output terminal the evolution of
+              results. Default: ``False``.
+        - workdir: Directory to write checkpoints. Default: execution
+              directory.
+        - checkpointing: A flag for checkpointing model state. Default:
+              ``False``. `RunTimeError` is generated if ``True`` and
+              tensorflow is not available.
+        - return_state: A flag for returning the train state instead of
+              the model variables. Default: ``False``, i.e. return model
+              variables.
+
+        Args:
+            config: Hyperparameter configuration.
+        """
+
+        # Determine logging configuration
+        if "log" in config:
+            self.logflag: bool = config["log"]
+            if self.logflag:
+                self.itstat_object, self.itstat_insert_func = stats_obj()
+        else:
+            self.logflag = False
+
+        # Determine checkpointing configuration
+        if "workdir" in config:
+            self.workdir: str = config["workdir"]
+        else:
+            self.workdir = "./"
+
+        if "checkpointing" in config:
+            self.checkpointing: bool = config["checkpointing"]
+        else:
+            self.checkpointing = False
+
+        # Determine variable to return at end of training
+        if "return_state" in config:
+            # Returning Flax train state
+            self.return_state = config["return_state"]
+        else:
+            # Return model variables
+            self.return_state = False
+
+    def configure_training_functions(self, config: ConfigDict):
+        """Construct training functions.
+
+        Default functions are used if not specified in configuration.
+
+        The functions constructed correspond to
+
+        - `create_lr_schedule`: A function that creates an Optax learning
```

Comment:
This is not rendering correctly in the docs for some reason.

Reply:
Fixed.